### DISCUSSION TITLE
Getting compiled IDL field names

### DISCUSSION THREAD

**Discussion by tmayoff** [OpenDDS/OpenDDS#3811] [Q&A] [answered]
Is there a way to get a list of a structs field names from an IDL, without creating a whole ValueWriter/Reader?

**Comment by mitza-oci** [comment, marked as answer]
There are a few options to consider.  
- If TAO's TypeCodes are enabled, you can use the TypeCode generated by tao_idl
- You can also use the XTypes TypeObject generated by opendds_idl
- The MetaStruct (internal type that OpenDDS uses for content-subscription features) has `getFieldNames`.  Search for that function name in `MetaStructTest.cpp` for an example

**Reply by iguessthislldo** [reply]
To add on: MetaStruct is probably the most straight forward. I was still trying to come up with an example for TypeObject, but it was taking too long to figure out. Here's the MetaStruct example I came up with that can be put in `DevGuideExamples/DCPS/Messenger`:

```C++
#include <dds/DCPS/FilterEvaluator.h>

// ...

      const OpenDDS::DCPS::MetaStruct& ms = OpenDDS::DCPS::getMetaStruct<Messenger::Message>();
      for (const char** name = ms.getFieldNames(); *name; name++) {
        printf("FIELD: %s\n", *name);
      }
```

Prints out:

```
FIELD: from
FIELD: subject
FIELD: subject_id
FIELD: text
FIELD: count
```

**Reply by mitza-oci** [reply]
For TypeObject this probably requires the "complete" form of the TypeObject, so enable that opendds_idl option.  Start with the TypeSupport for the IDL struct in question.  TypeSupportImpl has a method `getCompleteTypeIdentifier()`.  The resulting TypeIdentifier would be an input to `TypeLookupService::get_type_object`.  That type object is a CompleteTypeObject.  You can examine it directly or use `complete_to_dynamic` in the `TypeLookupService` to get a more user-friendly view.

**Reply by tmayoff** [reply]
Are there any downsides to any of the options, I got what I wanted working thanks to the above example, are there any reasons not to use the MetaStruct stuff?

**Reply by mitza-oci** [reply]
MetaStruct is internal to OpenDDS (not defined by any spec) so it may change at any time.

**Reply by tmayoff** [reply]
Are the other methods less likely to change?

**Reply by mitza-oci** [reply]
The ones based on the CORBA spec (TypeCode) and DDS-XTypes spec (TypeObject) are not likely to change.

**Reply by tmayoff** [reply]
Is it possible to use one of these methods at compile time using constexpr or something? This might be worth a new discussion but I'd like to be able to set the value of an IDL struct member only if it has that member without needed to add extra defines or preprocessor stuff (which is how I'm currently accomplishing it).

**Reply by mitza-oci** [reply]
If you're using a compiler with C++20 concepts that might be a good way to go

**Reply by iguessthislldo** [reply]
We don't have anything that would work in constexpr, at least as far as I know. I don't know how far C++ has gotten with introspection and reflection, but that might be worth looking into. Maybe Boost has something for that? Besides the IDL compiler, where we might do something like what you're describing, we do have [`DynamicDataAdapter`](https://github.com/OpenDDS/OpenDDS/blob/master/dds/DCPS/XTypes/DynamicDataAdapter.h). It allows accessing any C++ IDL value as a `DynamicData`, which allows it to be interacted with in an indirect way.

It was redesigned at the end of last year so it's now mostly functional, but the drawbacks listed in the header have kept it in the experimental side of things so it hasn't been officially announced. It should work just fine for most cases though. 

It needs [complete type objects to be generated](https://opendds.readthedocs.io/en/master/devguide/xtypes.html#enabling-use-of-completetypeobjects), but after that usage is pretty simple. Here's an example that sets a string member called "extra" on `message` that's the same type as TypeSupport `ts`, but only if that type has that member:

```C++
    const DDS::DynamicType_var dt = ts->get_type();
    const DDS::DynamicData_var dd = OpenDDS::XTypes::get_dynamic_data_adapter(dt, message);
    DDS::DynamicTypeMember_var member;
    dt->get_member_by_name(member, "extra");
    if (member) {
      dd->set_string_value(member->get_id(), "Hello");
    }
```

`MetaStruct` mentioned before could also do this, but it wasn't designed to be user friendly and we might eventually simplify the whole content-subscription system by replacing `MetaStruct` with `DynamicData`/`DynamicDataAdapter`.

**Reply by tmayoff** [reply]
I've not used concepts before but as it turns out they're dead simple.

```
template <typename T>
concept HasMachineID = requires(T t) { t.UID(); };
```

This was what I used on some overloaded functions and it works. I will however use the XTypes stuff for something else as I've been meaning to move away from the MetaStruct stuff that broke for me a little while ago.

Thank you both for the answers
